dan_layer/core/src/workers/pacemaker_worker.rs

```diff
@@ -0,0 +1,299 @@
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use std::{
+    collections::HashMap,
+    fmt::Debug,
+    hash::Hash,
+    sync::{Arc, Mutex},
+    time::Duration,
+};
+
+use log::*;
+use tari_shutdown::ShutdownSignal;
+use tokio::{
+    sync::{
+        mpsc::{Receiver, Sender},
+        oneshot,
+    },
+    task::JoinHandle,
+};
+
+use super::hotstuff_error::HotStuffError;
+
+const LOG_TARGET: &str = "tari::dan_layer::pacemaker_worker";
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum PacemakerWaitStatus {
+    WaitTimeOut,
+}
+
+async fn send_timeout_message<T: Debug + Send>(
+    wait_over: T,
+    tx_waiter_status: Sender<(T, PacemakerWaitStatus)>,
+) -> Result<(), HotStuffError> {
+    info!(target: LOG_TARGET, "Sending timeout message for value: {:?}", wait_over);
+    tx_waiter_status
+        .send((wait_over, PacemakerWaitStatus::WaitTimeOut))
+        .await
+        .map_err(|_| HotStuffError::SendError)?;
+    Ok(())
+}
+
+/// A pacemaker service that is responsible for:
+///   1. Receiving [`start_wait`] messages parametrized over a type `T` and start wait up to a max_timeout value.
+///   2. Receiving [`stop_wait`] messages for an already existing [`start_wait`].
+///   3. If it receives a stop message then it should stop the waiting and do nothing else;
+///   4. Otherwise, it should notify the sender of [`start_wait`] message that the max_timeout duration has passed.
+#[derive(Debug)]
+pub struct Pacemaker<T> {
+    /// Receiver for start signal. Whenever the service receives a new instance of `T`
+    /// it starts a waiting max_timeout time process
+    rx_start_signal: Receiver<T>,
+    /// Receiver of stop/shutdown signal. It is assumed that whenever a new value `T` is
+    /// received, that we are already waiting for timeout on that value (i.e. rx_start_signal,
+    /// received that same value first). If such value is received, then we stop the waiting process
+    rx_shutdown_signal: Receiver<T>,
+    /// An auxiliary map, mapping each received value to a single use channel (of buffer size 1).
+    /// Its role it to facilitate asynchronous communication between the arrival of shutdown messages
+    /// and time out signals
+    tx_inner_map: Arc<Mutex<HashMap<T, oneshot::Sender<()>>>>,
+    /// Sender which sends Timeout status to the other end of the channel
+    tx_waiter_status: Sender<(T, PacemakerWaitStatus)>,
+    /// Duration of timeout period
+    max_timeout: Duration,
+}
+
+impl<T: Clone + Debug + PartialEq + Eq + Hash + Send + Sync + 'static> Pacemaker<T> {
+    pub fn spawn(
+        rx_start_signal: Receiver<T>,
+        rx_shutdown_signal: Receiver<T>,
+        tx_waiter_status: Sender<(T, PacemakerWaitStatus)>,
+        max_timeout: Duration,
```

Comment:
Rename to this to something else. e.g. delta_time or latency, it will not be max_timeout for sure, because we will sometimes wait delta sometimes 2*delta

Reply:
So it is part of the overall message?  In that case, we need a check to see if the message contains an inner `Duration` type, so that we can use it to timeout

Reply:
What do you mean by inner `Duration`? The channel will be over `(T, Duration)`. The duration is not in the `T`